src/matrixfree/initialConditions.cc

```diff
@@ -296,6 +296,71 @@ MatrixFreePDE<dim, degree>::applyInitialConditions()
         }
     }
 
+  // 1) DETERMINE UNIQUE VTK FILES, CREATE DICTIONARY
+  std::unordered_map<std::string, std::vector<size_t>> file_field_map;
+  for (size_t i = 0; i < userInputs.number_of_variables; ++i) {
+      if (userInputs.load_ICs[i]){
+          file_field_map[userInputs.load_file_name[i]].push_back(i);
+      }
+  }
+  // Read out unique filenames
+  int proc_num = Utilities::MPI::this_mpi_process(MPI_COMM_WORLD);
+  if (proc_num == 0){
+      std::cout << "Unique VTK input files: " << std::endl;
+      for (std::unordered_map<std::string, std::vector<size_t>>::iterator it = file_field_map.begin();
+              it != file_field_map.end(); it++){
+          std::string filename = it->first;
+          std::cout << filename << ", ";
+      }
+  }
+  // 2) READ EACH VTK FILE ONCE, APPLY ICs FOR SCALAR FIELDS
+  typedef PRISMS::PField<double*, double, dim> ScalarField;
+  typedef PRISMS::Body<double*, dim> Body;
+  Body body;
+
+  for (std::unordered_map<std::string, std::vector<size_t>>::iterator it = file_field_map.begin();
+          it != file_field_map.end(); it++){
+      bool using_parallel_files = false;
+      std::string filename = it->first;
+      std::vector<size_t> index_list = it->second;
+      // For parallel file capability
+      for (size_t i = 0; i < index_list.size(); i++){
+          unsigned int var_index = index_list[i];
+          using_parallel_files = (using_parallel_files || userInputs.load_parallel_file[var_index]);
+      }
+      if (using_parallel_files){
+          std::ostringstream conversion;
+          conversion << proc_num;
+          filename = filename + "." + conversion.str() + ".vtk";
+      }
+      else {
+          filename = filename + ".vtk"; // add file extension
+      }
+      
+      std::cout << "Reading " << filename << "\n";
+      body.read_vtk(filename);
+
+      for(size_t i = 0; i < index_list.size(); i++){
+          unsigned int var_index = index_list[i];
+          std::string var_name = userInputs.load_field_name[var_index];
+
+          // Find the scalar field in the file
+          ScalarField &conc = body.find_scalar_field(var_name);
```

Comment:
Try using a more descriptive variable name.

Reply:
ScalarField is a class, not a variable.